tests/tt_eager/python_api_testing/unit_testing/misc/test_all_gather.py

```diff
@@ -435,94 +435,235 @@ def test_all_gather_on_t3000_nightly(
+    logger.info(f"tensor_layout: {tensor_layout}")
+    logger.info(f"tensor_mem_layout: {tensor_mem_layout}")
+    logger.info(f"orientation: {orientation}")
+    # logger.info(f"num_cores: {num_cores}")
+    logger.info(f"shard_grid: {shard_grid}")
+    logger.info(f"input_shard_shape: {input_shard_shape}")
+
+    # TODO: Figure out better shard spec
+    # all_cores_.size() == 1
+    # this->num_cores() == 8
+    # shard_grid = ttl.tensor.CoreRangeSet(ttl.tensor.num_cores_to_corerange_set(num_cores, compute_grid_size, True))
 
-    input_tensors = torch.chunk(input_tensor, num_devices, dim)
-    tt_input_tensors = []
-    compute_grid_size = devices[0].compute_with_storage_grid_size()
-    shard_grid = ttl.tensor.CoreRangeSet(ttl.tensor.num_cores_to_corerange_set(num_cores, compute_grid_size, True))
     input_shard_spec = ttl.tensor.ShardSpec(
         shard_grid,
-        [
-            input_tensors[0].shape.numel() // input_tensors[0].shape[-1],
-            input_tensors[0].shape[-1] // num_cores,
-        ],
-        ttl.tensor.ShardOrientation.ROW_MAJOR,
+        input_shard_shape,
+        orientation,
         False,
     )
     input_mem_config = ttl.tensor.MemoryConfig(
-        ttl.tensor.TensorMemoryLayout.WIDTH_SHARDED, ttl.tensor.BufferType.L1, input_shard_spec
+        tensor_mem_layout, buffer_type=ttl.tensor.BufferType.L1, shard_spec=input_shard_spec
     )
+    output_shard_shape = list(input_shard_shape)
+    if dim == 3:
+        output_shard_shape[1] *= num_devices
+    else:
+        output_shard_shape[0] *= num_devices
     output_shard_spec = ttl.tensor.ShardSpec(
         shard_grid,
-        [
-            input_tensors[0].shape.numel() // input_tensors[0].shape[-1],
-            input_shape[-1] // num_cores,
-        ],
-        ttl.tensor.ShardOrientation.ROW_MAJOR,
+        output_shard_shape,
+        orientation,
         False,
     )
     output_mem_config = ttl.tensor.MemoryConfig(
-        ttl.tensor.TensorMemoryLayout.WIDTH_SHARDED, ttl.tensor.BufferType.L1, output_shard_spec
+        tensor_mem_layout, buffer_type=ttl.tensor.BufferType.L1, shard_spec=output_shard_spec
     )
+
+    if num_devices < 2:
+        pytest.skip("Requires multiple devices to run")
+    elif num_devices == 2 and num_links == 2:
+        pytest.skip("Not enough links to run")
+
+    if unchunked_input_shape[dim] % num_devices != 0 or (
+        dim == 3 and unchunked_input_shape[dim] // num_devices % 32 != 0
+    ):
+        pytest.skip("Unsupported test case")
+
+    tt_input_tensors = []
+
     for i, t in enumerate(input_tensors):
-        tt_input_tensors.append(ttl.tensor.Tensor(t, input_dtype).to(layout).to(devices[i], input_mem_config))
+        tt_input_tensors.append(ttl.tensor.Tensor(t, input_dtype).to(tensor_layout).to(devices[i], input_mem_config))
 
     tt_out_tensors = ttl.tensor.all_gather(tt_input_tensors, dim, num_links, output_mem_config=output_mem_config)
+    for d in devices:
```

Comment:
is this explicitly needed?

Reply:
technically it's not needed no, but there was some general instability with synchronizing between calls (other issues are open). Ideally we shouldn't need it (unless we wanted to see iteration by iteration progress.